cumulus/primitives/parachain-inherent/src/lib.rs

```diff
@@ -85,6 +88,276 @@ pub mod v0 {
 	}
 }
 
+/// An inbound message whose content was hashed.
+#[derive(
+	codec::Encode,
+	codec::Decode,
+	codec::DecodeWithMemTracking,
+	sp_core::RuntimeDebug,
+	Clone,
+	PartialEq,
+	TypeInfo,
+)]
+pub struct HashedMessage<BlockNumber = RelayChainBlockNumber> {
+	pub sent_at: BlockNumber,
+	pub msg_hash: sp_core::H256,
+}
+
+impl<BlockNumber: Copy> From<&InboundDownwardMessage<BlockNumber>> for HashedMessage<BlockNumber> {
+	fn from(msg: &InboundDownwardMessage<BlockNumber>) -> Self {
+		Self { sent_at: msg.sent_at, msg_hash: MessageQueueChain::hash_msg(&msg.msg) }
+	}
+}
+
+impl From<&InboundHrmpMessage> for HashedMessage<RelayChainBlockNumber> {
+	fn from(msg: &InboundHrmpMessage) -> Self {
+		Self { sent_at: msg.sent_at, msg_hash: MessageQueueChain::hash_msg(&msg.data) }
+	}
+}
+
+pub trait InboundMessage<BlockNumber> {
+	type CompressedMessage: Debug;
+
+	fn data(&self) -> &[u8];
+
+	fn sent_at(&self) -> BlockNumber;
+
+	fn to_compressed(&self) -> Self::CompressedMessage;
+}
+
+/// A collection of inbound messages.
+#[derive(
+	codec::Encode,
+	codec::Decode,
+	codec::DecodeWithMemTracking,
+	sp_core::RuntimeDebug,
+	Clone,
+	PartialEq,
+	TypeInfo,
+)]
+pub struct InboundMessagesCollection<BlockNumber, Message: InboundMessage<BlockNumber>> {
+	messages: Vec<Message>,
+
+	_phantom: core::marker::PhantomData<BlockNumber>,
+}
+
+impl<BlockNumber: PartialEq + PartialOrd, Message: InboundMessage<BlockNumber>>
+	InboundMessagesCollection<BlockNumber, Message>
+{
+	pub fn new(messages: Vec<Message>) -> Self {
+		Self { messages, _phantom: Default::default() }
+	}
+
+	/// Drop all the messages up to `last_processed_msg`.
+	pub fn drop_processed_messages(&mut self, last_processed_msg: &InboundMessageId<BlockNumber>) {
+		let mut last_processed_msg_idx = None;
+		let messages = &mut self.messages;
+		for (rev_idx, message) in messages.iter().rev().enumerate() {
```

Comment:
I'm confused why we need here the reversed index etc? Could we not just take the index of the last message + 1?